src/contextLogger.js

```diff
@@ -0,0 +1,134 @@
+
+ Contributors
+ --------------
+ This is the official list of the Mojaloop project contributors for this file.
+ Names of the original copyright holders (individuals or organizations)
+ should be listed with a '*' in the first column. People who have
+ contributed from an organization can be listed under the organization
+ that actually holds the copyright for their contributions (see the
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ * Eugen Klymniuk <eugen.klymniuk@infitx.com>
+ --------------
+ **********/
+
+/* eslint-disable space-before-function-paren */
+const { AsyncLocalStorage } = require('node:async_hooks')
+const stringify = require('safe-stable-stringify')
+const mlLogger = require('./index')
+
+const asyncStorage = new AsyncLocalStorage()
+
+const loggerFactory = (context = null) => new ContextLogger(context)
+
+class ContextLogger {
+  constructor (context) {
+    this.mlLogger = mlLogger
+    this.context = this.createContext(context)
+
+    this.isErrorEnabled = this.mlLogger.isLevelEnabled('error')
+    this.isWarnEnabled = this.mlLogger.isLevelEnabled('warn')
+    this.isAuditEnabled = this.mlLogger.isLevelEnabled('audit')
+    this.isTraceEnabled = this.mlLogger.isLevelEnabled('trace')
+    this.isInfoEnabled = this.mlLogger.isLevelEnabled('info')
+    this.isPerfEnabled = this.mlLogger.isLevelEnabled('perf')
+    this.isVerboseEnabled = this.mlLogger.isLevelEnabled('verbose')
+    this.isDebugEnabled = this.mlLogger.isLevelEnabled('debug')
+    this.isSillyEnabled = this.mlLogger.isLevelEnabled('silly')
+  }
+
+  error(message, meta) {
+    this.isErrorEnabled && this.mlLogger.error(this.formatLog(message, meta))
+  }
+
+  warn(message, meta) {
+    this.isWarnEnabled && this.mlLogger.warn(this.formatLog(message, meta))
+  }
+
+  info(message, meta) {
+    this.isInfoEnabled && this.mlLogger.info(this.formatLog(message, meta))
+  }
+
+  verbose(message, meta) {
+    this.isVerboseEnabled && this.mlLogger.verbose(this.formatLog(message, meta))
+  }
+
+  debug(message, meta) {
+    this.isDebugEnabled && this.mlLogger.debug(this.formatLog(message, meta))
+  }
+
+  silly(message, meta) {
+    this.isSillyEnabled() && this.mlLogger.silly(this.formatLog(message, meta))
+  }
+
+  audit (message, meta) {
+    this.isAuditEnabled && this.mlLogger.audit(this.formatLog(message, meta))
+  }
+
+  trace(message, meta) {
+    this.isTraceEnabled && this.mlLogger.trace(this.formatLog(message, meta))
+  }
+
+  perf(message, meta) {
+    this.isPerfEnabled && this.mlLogger.perf(this.formatLog(message, meta))
+  }
```

Comment:
I have one proposal around logging, which is not mandatory but is a small improvement, related to performance.
It is a very common case where log methods are invoked with message being assembled by caller, for example:
```js
if (logger.isInfoEnabled) logger.info(`Error attempting HTTP ${method} statusCode: ${statusCode}`)
```

If we move the check for if the level is enabled in the method, the string will need to be constructed and passed to the method, only to discover that the log level may be turned off and then string must be garbage collected. This can be undesirable, especially for trace level, where many strings may undergo the same processing. I would suggest the following shorter and optimal pattern:

```js
logger.info?.(`Error attempting HTTP ${method} statusCode: ${statusCode}`)
```

It avoids the problem and the check for if a level is enabled is done by checking if the function is defined.
I will approve the PR, but if you feel good about following the above pattern, the code above must be refactored.

Reply:
First of all, I'd suggest avoiding using template strings for log message (1st arg of logMethod). Instead, it's better to use meta - the 2nd arg:
```
logger.info('Error attempting send HTTP:', { statusCode, method })
```
But in case if somebody wants to use template string, I don't mind that optimization. The only thing is I didn't get how it works.  In your example with `logger.info?.(...)`  _.info()_ method will be called in any case, regardless of log level. Or I'm wrong?

Reply:
It will only be called if not null or undefined. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining, which also works for function calling.

So to turning off a level we need need to not define the function for that level.
Templates for error messages are good in cases we want to have user friendly messages, where parameters are interpolated in the middle of the message.

Reply:
> ...user friendly messages

but logMessage is not for user )  it's for us (developers), and to get more details about context of process execution in that particular place in code, we have second param - meta